File: flutter/rum/lib/screens/home_screen.dart
import 'package:flutter/material.dart';
import 'package:last9_rum_flutter/last9_rum_flutter.dart';

import '../api.dart';
import '../config.dart';
import '../event_log.dart';
import '../theme.dart';
import '../widgets.dart';

/// Home tab — owns its own [Navigator] so view tracking fires on the
/// Dashboard → Detail push (the root [L9NavigationObserver] sees each route).
class HomeScreen extends StatelessWidget {
  const HomeScreen({super.key});

  @override
  Widget build(BuildContext context) {
    return Navigator(
      observers: <NavigatorObserver>[L9NavigationObserver()],
      onGenerateRoute: (RouteSettings settings) {
        if (settings.name == DetailScreen.routeName) {
          final int postId = (settings.arguments as int?) ?? 1;
          return MaterialPageRoute<void>(
            settings: settings,
            builder: (_) => DetailScreen(postId: postId),
          );
        }
        return MaterialPageRoute<void>(
          settings: const RouteSettings(name: DashboardScreen.routeName),
          builder: (_) => const DashboardScreen(),
        );
      },
    );
  }
}

class DashboardScreen extends StatefulWidget {
  const DashboardScreen({super.key});

  static const String routeName = 'Home';

  @override
  State<DashboardScreen> createState() => _DashboardScreenState();
}

class _DashboardScreenState extends State<DashboardScreen> {
  List<dynamic> _posts = <dynamic>[];
  List<dynamic> _users = <dynamic>[];
  List<dynamic> _comments = <dynamic>[];
  List<ApiResult> _homeRequests = <ApiResult>[];
  bool _loading = true;

  @override
  void initState() {
    super.initState();
    _loadHomeData();
  }

  Future<void> _loadHomeData() async {
    setState(() => _loading = true);
    await L9Rum.startView('Home');
    addLog('startView: Home');
    try {
      final List<TimedJsonResult<dynamic>> results =
          await Future.wait<TimedJsonResult<dynamic>>(<Future<TimedJsonResult<dynamic>>>[
        timedJson<List<dynamic>>('Home fast posts', '$kApiBase/posts?_limit=20',
            const DemoRequestTags(tab: 'home', name: 'posts-list')),
        timedJson<List<dynamic>>('Home fast users', '$kApiBase/users?_limit=5',
            const DemoRequestTags(tab: 'home', name: 'users-list')),
        timedJson<List<dynamic>>('Home fast comments',
            '$kApiBase/comments?postId=1',
            const DemoRequestTags(tab: 'home', name: 'comments-for-post')),
        timedJson<Map<String, dynamic>>('Home delayed 1s',
            'https://httpbin.org/delay/1',
            const DemoRequestTags(tab: 'home', name: 'delay-1s')),
        timedJson<Map<String, dynamic>>('Home delayed 3s',
            'https://httpbin.org/delay/3',
            const DemoRequestTags(tab: 'home', name: 'delay-3s')),
      ]);

      final List<ApiResult> requestResults =
          results.map((TimedJsonResult<dynamic> r) => r.result).toList();
      final ApiResult maxResult = requestResults.reduce(
        (ApiResult a, ApiResult b) => b.durationMs > a.durationMs ? b : a,
      );
      if (!mounted) return;
      setState(() {
        _posts = (results[0].data as List<dynamic>?) ?? <dynamic>[];
        _users = (results[1].data as List<dynamic>?) ?? <dynamic>[];
        _comments = (results[2].data as List<dynamic>?) ?? <dynamic>[];
        _homeRequests = requestResults;
      });
      addLog(
        'Home APIs complete; expected max view.ttfd source: '
        '${maxResult.label} (${maxResult.durationMs}ms)',
      );
    } catch (e) {
      addLog('Home APIs failed: $e');
    } finally {
      if (mounted) setState(() => _loading = false);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Posts')),
      body: ListView(
        padding: const EdgeInsets.all(16),
        children: <Widget>[
          if (_loading)
            AppCard(
              child: Column(
                children: const <Widget>[
                  Padding(
                    padding: EdgeInsets.all(20),
                    child: CircularProgressIndicator(),
                  ),
                  Hint(
                    'Loading posts, users, and comments before the Home '
                    'screen is fully displayed.',
                  ),
                ],
              ),
            )
          else ...<Widget>[
            const FeatureBadge(features: <String>[
              'Home starts an active View before API requests',
              'Fast and delayed GET requests run before full content render',
              'SDK sets view.ttfd from the maximum request time on this view',
              'Home APIs include l9_demo_tab=home query tags',
            ]),
            SummaryCard(
              title: 'Home full display data',
              lines: <String>[
                '${_posts.length} posts',
                '${_users.length} users',
                '${_comments.length} comments for the featured post',
              ],
            ),
            const SectionTitle('TTFD Request Timings'),
            const Hint(
              'The delayed 3s request should usually be the max-duration '
              'source for view.ttfd. Filter dashboard URLs by '
              'l9_demo_tab=home or l9_demo_request=delay-3s.',
            ),
            for (final ApiResult r in _homeRequests) ApiResultCard(result: r),
            const SizedBox(height: 8),
            for (final dynamic post in _posts)
              _PostTile(
                post: post,
                onTap: () {
                  final int id = (post['id'] as num?)?.toInt() ?? 1;
                  L9Rum.addEvent('nav_tap',
                      attributes: <String, dynamic>{'destination': 'Post #$id'});
                  addLog('nav → Post #$id');
                  Navigator.of(context)
                      .pushNamed(DetailScreen.routeName, arguments: id);
                },
              ),
            const SectionTitle('Featured Users'),
            for (final dynamic user in _users)
              AppCard(
                radius: 8,
                leftBorderColor: AppColors.accent,
                leftBorderWidth: 3,
                padding: const EdgeInsets.all(12),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: <Widget>[
                    Text((user['name'] ?? '').toString(),
                        style: const TextStyle(
                            fontSize: 13,
                            fontWeight: FontWeight.w600,
                            color: AppColors.textPrimary)),
                    Text((user['email'] ?? '').toString(),
                        style: const TextStyle(
                            fontSize: 11, color: AppColors.textMuted)),
                  ],
                ),
              ),
            const SectionTitle('Featured Comments'),
            for (final dynamic c in _comments.take(3))
              AppCard(
                radius: 8,
                leftBorderColor: AppColors.ok,
                leftBorderWidth: 3,
                padding: const EdgeInsets.all(12),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: <Widget>[
                    Text((c['name'] ?? '').toString(),
                        style: const TextStyle(
                            fontSize: 13,
                            fontWeight: FontWeight.w600,
                            color: AppColors.textPrimary)),
                    Text((c['body'] ?? '').toString(),
                        maxLines: 2,
                        overflow: TextOverflow.ellipsis,
                        style: const TextStyle(
                            fontSize: 10,
                            color: AppColors.textMuted,
                            fontFamily: 'monospace')),
                  ],
                ),
              ),
            const SizedBox(height: 16),
          ],
        ],
      ),
    );
  }
}

class _PostTile extends StatelessWidget {
  const _PostTile({required this.post, required this.onTap});

  final dynamic post;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    return AppCard(
      radius: 10,
      padding: const EdgeInsets.all(14),
      onTap: onTap,
      child: Row(
        children: <Widget>[
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: <Widget>[
                Text(
                  (post['title'] ?? '').toString(),
                  maxLines: 1,
                  overflow: TextOverflow.ellipsis,
                  style: const TextStyle(
                      fontSize: 14,
                      fontWeight: FontWeight.w600,
                      color: AppColors.textPrimary),
                ),
                const SizedBox(height: 2),
                Text(
                  (post['body'] ?? '').toString(),
                  maxLines: 1,
                  overflow: TextOverflow.ellipsis,
                  style:
                      const TextStyle(fontSize: 11, color: AppColors.textMuted),
                ),
              ],
            ),
          ),
          const Text('›',
              style: TextStyle(fontSize: 20, color: Color(0xFFCCCCCC))),
        ],
      ),
    );
  }
}

class DetailScreen extends StatefulWidget {
  const DetailScreen({super.key, required this.postId});

  static const String routeName = 'Detail';

  final int postId;

  @override
  State<DetailScreen> createState() => _DetailScreenState();
}

class _DetailScreenState extends State<DetailScreen> {
  Map<String, dynamic>? _post;
  List<dynamic> _comments = <dynamic>[];
  bool _loading = true;

  @override
  void initState() {
    super.initState();
    _load();
  }

  Future<void> _load() async {
    final int id = widget.postId;
    const DemoRequestTags postTags =
        DemoRequestTags(tab: 'home', name: 'detail-post');
    const DemoRequestTags commentsTags =
        DemoRequestTags(tab: 'home', name: 'detail-comments');
    final List<TimedJsonResult<dynamic>> results =
        await Future.wait<TimedJsonResult<dynamic>>(<Future<TimedJsonResult<dynamic>>>[
      timedJson<Map<String, dynamic>>(
          'Detail post', '$kApiBase/posts/$id', postTags),
      timedJson<List<dynamic>>(
          'Detail comments', '$kApiBase/posts/$id/comments', commentsTags),
    ]);
    if (!mounted) return;
    setState(() {
      _post = results[0].data as Map<String, dynamic>?;
      _comments = (results[1].data as List<dynamic>?) ?? <dynamic>[];
      _loading = false;
    });
    addLog('GET /posts/$id → ${results[0].result.status}');
    addLog(
        'GET /posts/$id/comments → ${results[1].result.status} (${_comments.length})');
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        leading: TextButton(
          onPressed: () => Navigator.of(context).maybePop(),
          child: const Text('← Back',
              style: TextStyle(
                  fontSize: 14,
                  color: AppColors.accent,
                  fontWeight: FontWeight.w600)),
        ),
        leadingWidth: 80,
        title: Text('Post #${widget.postId}'),
      ),
      body: _loading
          ? const Center(
              child: Padding(
                  padding: EdgeInsets.all(20),
                  child: CircularProgressIndicator()))
          : ListView(
              padding: const EdgeInsets.all(16),
              children: <Widget>[
                SectionTitle((_post?['title'] ?? '').toString()),
                Padding(
                  padding: const EdgeInsets.only(bottom: 16),
                  child: Text((_post?['body'] ?? '').toString(),
                      style: AppText.hint),
                ),
                SectionTitle('Comments (${_comments.length})'),
                for (final dynamic c in _comments)
                  AppCard(
                    radius: 8,
                    leftBorderColor: AppColors.accent,
                    leftBorderWidth: 3,
                    padding: const EdgeInsets.all(12),
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: <Widget>[
                        Text((c['name'] ?? '').toString(),
                            style: const TextStyle(
                                fontSize: 13,
                                fontWeight: FontWeight.w600,
                                color: AppColors.textPrimary)),
                        Text((c['email'] ?? '').toString(),
                            style: const TextStyle(
                                fontSize: 11, color: AppColors.textMuted)),
                        const SizedBox(height: 4),
                        Text((c['body'] ?? '').toString(),
                            style: const TextStyle(
                                fontSize: 10,
                                color: AppColors.textMuted,
                                fontFamily: 'monospace')),
                      ],
                    ),
                  ),
                const SizedBox(height: 16),
              ],
            ),
    );
  }
}

File: android/app/src/main/java/io/last9/rumexample/screens/HomeScreen.kt
package io.last9.rumexample.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.navigation.NavType
import androidx.navigation.compose.NavHost
import androidx.navigation.compose.composable
import androidx.navigation.compose.rememberNavController
import androidx.navigation.navArgument
import io.last9.rum.L9Rum
import io.last9.rumexample.EventLog
import io.last9.rumexample.data.Api
import io.last9.rumexample.data.ApiResult
import io.last9.rumexample.data.Comment
import io.last9.rumexample.data.DemoRequestTags
import io.last9.rumexample.data.Post
import io.last9.rumexample.data.User
import io.last9.rumexample.data.parseComments
import io.last9.rumexample.data.parsePost
import io.last9.rumexample.data.parsePosts
import io.last9.rumexample.data.parseUsers
import io.last9.rumexample.ui.AccentEntryCard
import io.last9.rumexample.ui.ApiResultCard
import io.last9.rumexample.ui.FeatureBadge
import io.last9.rumexample.ui.Hint
import io.last9.rumexample.ui.L9Theme
import io.last9.rumexample.ui.LoadingCard
import io.last9.rumexample.ui.PostListItem
import io.last9.rumexample.ui.ScreenHeader
import io.last9.rumexample.ui.SectionTitle
import io.last9.rumexample.ui.SummaryCard

private object HomeRoutes {
    const val DASHBOARD = "dashboard"
    const val DETAIL = "detail/{postId}"
    fun detail(postId: Int) = "detail/$postId"
}

/**
 * Home tab — a nested navigation stack (Dashboard → Detail) using real Compose
 * navigation routes, mirroring the reference app's Home stack.
 */
@Composable
fun HomeScreen() {
    val nav = rememberNavController()
    NavHost(navController = nav, startDestination = HomeRoutes.DASHBOARD) {
        composable(HomeRoutes.DASHBOARD) {
            DashboardScreen(onOpenPost = { postId ->
                L9Rum.addEvent("nav_tap", mapOf("destination" to "Post #$postId"))
                EventLog.add("nav → Post #$postId")
                nav.navigate(HomeRoutes.detail(postId))
            })
        }
        composable(
            HomeRoutes.DETAIL,
            arguments = listOf(navArgument("postId") { type = NavType.IntType }),
        ) { entry ->
            val postId = entry.arguments?.getInt("postId") ?: 1
            DetailScreen(postId = postId, onBack = { nav.popBackStack() })
        }
    }
}

@Composable
private fun DashboardScreen(onOpenPost: (Int) -> Unit) {
    val context = LocalContext.current
    val posts = remember { mutableStateListOf<Post>() }
    val users = remember { mutableStateListOf<User>() }
    val comments = remember { mutableStateListOf<Comment>() }
    val homeRequests = remember { mutableStateListOf<ApiResult>() }
    var loading by remember { mutableStateOf(true) }

    LaunchedEffect(Unit) {
        loading = true
        L9Rum.startView("Home")
        EventLog.add("startView: Home")
        // Five requests in parallel (instrumented OkHttp), off the main thread.
        val postsRes = Api.timedGet(context, "Home fast posts", "https://jsonplaceholder.typicode.com/posts?_limit=20", DemoRequestTags("home", "posts-list"))
        val usersRes = Api.timedGet(context, "Home fast users", "https://jsonplaceholder.typicode.com/users?_limit=5", DemoRequestTags("home", "users-list"))
        val commentsRes = Api.timedGet(context, "Home fast comments", "https://jsonplaceholder.typicode.com/comments?postId=1", DemoRequestTags("home", "comments-for-post"))
        val delayOneRes = Api.timedGet(context, "Home delayed 1s", "https://httpbin.org/delay/1", DemoRequestTags("home", "delay-1s"))
        val delayThreeRes = Api.timedGet(context, "Home delayed 3s", "https://httpbin.org/delay/3", DemoRequestTags("home", "delay-3s"))

        val requestResults = listOf(
            postsRes.second, usersRes.second, commentsRes.second,
            delayOneRes.second, delayThreeRes.second,
        )
        posts.clear(); posts.addAll(parsePosts(postsRes.first))
        users.clear(); users.addAll(parseUsers(usersRes.first))
        comments.clear(); comments.addAll(parseComments(commentsRes.first))
        homeRequests.clear(); homeRequests.addAll(requestResults)
        val maxResult = requestResults.maxByOrNull { it.durationMs }
        EventLog.add("Home APIs complete; expected max view.ttfd source: ${maxResult?.label} (${maxResult?.durationMs}ms)")
        loading = false
    }

    Column(modifier = Modifier.fillMaxSize().background(L9Theme.ScreenBg)) {
        ScreenHeader("Posts")
        Column(
            modifier = Modifier.fillMaxSize().verticalScroll(rememberScrollState()).padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            if (loading) {
                LoadingCard("Loading posts, users, and comments before the Home screen is fully displayed.")
            } else {
                FeatureBadge(
                    features = listOf(
                        "Home starts an active View before API requests",
                        "Fast and delayed GET requests run before full content render",
                        "SDK sets view.ttfd from the maximum request time on this view",
                        "Home APIs include l9_demo_tab=home query tags",
                    ),
                )
                SummaryCard(
                    title = "Home full display data",
                    lines = listOf(
                        "${posts.size} posts",
                        "${users.size} users",
                        "${comments.size} comments for the featured post",
                    ),
                )
                SectionTitle("TTFD Request Timings")
                Hint("The delayed 3s request should usually be the max-duration source for view.ttfd. Filter dashboard URLs by l9_demo_tab=home or l9_demo_request=delay-3s.")
                homeRequests.forEach { r ->
                    ApiResultCard(r.label, r.status, r.ok, r.durationMs, r.error)
                }
                posts.forEach { post ->
                    PostListItem(post.title, post.body, onClick = { onOpenPost(post.id) })
                }
                SectionTitle("Featured Users")
                users.forEach { u ->
                    AccentEntryCard(accent = L9Theme.Accent, label = u.name, sub = u.email)
                }
                SectionTitle("Featured Comments")
                comments.take(3).forEach { c ->
                    AccentEntryCard(accent = L9Theme.Ok, label = c.name, body = c.body)
                }
            }
            Spacer(Modifier.size(8.dp))
        }
    }
}

@Composable
private fun DetailScreen(postId: Int, onBack: () -> Unit) {
    val context = LocalContext.current
    var post by remember { mutableStateOf<Post?>(null) }
    val comments = remember { mutableStateListOf<Comment>() }
    var loading by remember { mutableStateOf(true) }

    LaunchedEffect(postId) {
        loading = true
        val postRes = Api.timedGet(context, "detail-post", "https://jsonplaceholder.typicode.com/posts/$postId", DemoRequestTags("home", "detail-post"))
        val commentsRes = Api.timedGet(context, "detail-comments", "https://jsonplaceholder.typicode.com/posts/$postId/comments", DemoRequestTags("home", "detail-comments"))
        post = parsePost(postRes.first)
        val parsed = parseComments(commentsRes.first)
        comments.clear(); comments.addAll(parsed)
        loading = false
        EventLog.add("GET /posts/$postId → ${postRes.second.status}")
        EventLog.add("GET /posts/$postId/comments → ${commentsRes.second.status} (${parsed.size})")
    }

    Column(modifier = Modifier.fillMaxSize().background(L9Theme.ScreenBg)) {
        ScreenHeader("Post #$postId", onBack = onBack)
        Column(
            modifier = Modifier.fillMaxSize().verticalScroll(rememberScrollState()).padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            if (loading) {
                LoadingCard()
            } else {
                SectionTitle(post?.title ?: "")
                Hint(post?.body ?: "")
                SectionTitle("Comments (${comments.size})")
                comments.forEach { c ->
                    AccentEntryCard(accent = L9Theme.Accent, label = c.name, sub = c.email, body = c.body)
                }
            }
            Spacer(Modifier.size(8.dp))
        }
    }
}
